accepted/00062-hub-xmlrpc-api.md

```diff
@@ -0,0 +1,129 @@
+  * Get the `serverId` for a given client (based on `minion_id`, FQDN or other unique identifier. Normal client IDs will not work, as they will overlap between Servers)
+
+### Gateway functionality
+
+Idea is to expose existing Server XMLRPC endpoints/methods as they are on the Hub. The Hub will simply route calls to the right Server(s) and relay back results, rather than reimplementing any functionality. Areas are detailed below.
+
+* Authorization and authentication
+  * Gateway API will be secured via `hubSessionKey` tokens analoguously to Server's API
+    * Actually, Gateway will delegate authentication to the Hub API
+  * Consuming a Server's API from the Gateway continues to require a valid Server `sessionKey` token
+  * There are three ways to attach Server sessions to the Hub session
+    * Manual mode: programmer "attaches" Server sessions explicitly to his Hub session. Attachment lasts until logout
+      * `hub.login(username, password)` → `hubSessionKey`
+      * `hub.attachToServer(hubSessionKey, serverId, username, password)` → `serverSessionKey`
+      * `hub.serverMethod(serverSessionKey, parameters)` → `output`
+      * N+1 pairs of credentials will be needed (one for the Hub, one per each of the N Servers)
+     * Authentication relay mode: as an additional convenience option, which can be activated with a flag, Hub-User credentials can be re-used to authenticate against Servers. So if a certain username/password is uniformly configured on the Hub and across several Servers, Gateway usage becomes simpler: only one log in is needed, with its Hub-User credentials
+     * Automatic connect mode: as an additional convenience option, which can be activated with a flag, the Hub gateway API can automatically select the list of Servers a Hub-User works with, based on the list of Servers that same Hub-User has permissions on
+       - note: that this only makes sense if users are OK with the idea of using Server sysadmin users to act on Clients
+       - If they prefer to manage Clients via users that have no permissions to manage Servers, they can create a Hub-Org with no Servers. Any Hub-User in this Hub-Org will be able to connect to the Gateway, and possibly have permissions on Clients (if Servers have Server-Users with same name and passwords)
+
+* Multicasting: call an XMLRPC method on _N_ Servers at once
+  * method name stays the same
+  * Assuming the method on the Server accepts _M_ parameters, the same method on the Hub will accept an array of _N_ parameters, each of which contains as elements the _M_ parameters for a given Server
+    * eg. `actionchain.listChains(sessionKey)` → `hub.actionchain.listChains([serverSessionKey1, serverSessionKey2, ...])`
+    * eg. `channel.org.enableAccess(sessionKey, channelLabel, orgId)` → `hub.channel.org.enableAccess([serverSessionKey1, channelLabel1, orgId1], [serverSessionKey2, channelLabel2, orgId2], ...)`
+  * return types would be arrays of _N_ elements
```

Comment:
So the server's ids aren't required in this scenario? Are they automatically detected by identifying the matching `serverSessionKey`?
Just asking: Should we also add a "convenience feature" for calls that are exactly the same for all servers, where I only provide the dict of serverSessionKey's and the rest of the parameters only once?
Not a big deal because one certainly can expand those in the calling (e.g. Python) code as well.

Reply:
> So the server's ids aren't required in this scenario? Are they automatically detected by identifying the matching serverSessionKey?

That's the idea in this example, yes.

>  Should we also add a "convenience feature" for calls that are exactly the same for all servers, where I only provide the dict of serverSessionKey's and the rest of the parameters only once?

Yeah, why not. Added in the next steps.